Mage/src/main/java/mage/abilities/effects/common/HideawayPlayEffect.java

```diff
@@ -38,33 +43,41 @@ public boolean apply(Game game, Ability source) {
         if (controller == null || zone == null || zone.isEmpty()) {
             return true;
         }
-        Card card = zone.getRandom(game);
-        if (card == null) {
-            return false;
-        }
-        if (!controller.chooseUse(Outcome.PlayForFree, "Play " + card.getIdName() + " for free?", source, game)) {
-            return false;
-        }
-        card.setFaceDown(false, game);
-        int zcc = card.getZoneChangeCounter(game);
+        CardsImpl cards = new CardsImpl(zone.getCards(game));
+
+        boolean cardPlayed = false;
+        TargetCard target = new TargetCard(0, cards.size(), Zone.EXILED, new FilterCard("cards to play"));
+        controller.choose(Outcome.PlayForFree, cards, target, source, game);
+        List<UUID> targets = target.getTargets();
 
-        /* 702.74. Hideaway, rulings:
-         * If the removed card is a land, you may play it as a result of the last ability only if it's your turn
-         * and you haven't already played a land that turn. This counts as your land play for the turn.
-         * TODO: this doesn't work correctly with the back half of MDFCs
-         */
-        if (card.isLand(game)) {
-            UUID playerId = controller.getId();
-            if (!game.isActivePlayer(playerId) || !game.getPlayer(playerId).canPlayLand()) {
-                return false;
+        for (UUID targetId : targets) {
+            Card card = game.getCard(targetId);
+            /* 702.74. Hideaway, rulings:
+             * If the removed card is a land, you may play it as a result of the last ability only if it's your turn
+             * and you haven't already played a land that turn. This counts as your land play for the turn.
+             * TODO: this doesn't work correctly with the back half of MDFCs
+             */
+            if (card.isLand(game)) {
+                UUID playerId = controller.getId();
+                if (!game.isActivePlayer(playerId) || !game.getPlayer(playerId).canPlayLand()) {
+                    continue;
+                }
             }
-        }
 
-        if (!controller.playCard(card, game, true, new ApprovingObject(source, game))) {
-            if (card.getZoneChangeCounter(game) == zcc) {
-                card.setFaceDown(true, game);
+            if (!controller.chooseUse(Outcome.PlayForFree, "Play " + card.getIdName() + " for free?", source, game)) {
```

Comment:
1. You can filter compatible cards before any choices (e.g. remove bad lands from a playable list);
2. No needs in yes/no dialogs cause user already choose cards to cast (also chosen targets returned in chosen order);

Reply:
1. Is there a good way to make it do "you can target any number of nonlands but only up to the number of lands you can still play this turn"? I'm open to doing that if there's an easy way to do that but otherwise, seems simpler to let them target each thing and then just not attempt to play extra lands.
2. I copied this flow from [[Arcane Bombardment]], should I also make that try to cast everything that the player targeted rather than target and then prompt? (And yes, the chosen order of targets is precisely why I made this use targeting rather than just iterating through the `Set` returned by `zone.getCards(game)` because we need to allow players to choose the order in which the cards are played.)
3. Bonus question: Should I add support for [[Evercoat Ursine]] (I'm thinking an optional boolean parameter 'playOne' which sets the max targets to 1 on the TargetCard, since Evercoat Ursine's hideaway play effect was made knowing that you would hideaway multiple times) while I'm refactoring this ability anyways?

Reply:
> 1. Is there a good way to make it do "you can target any number of nonlands but only up to the number of lands you can still play this turn"?

Nope, you can't find precise amount before cast. Some effects can allow to play additional lands and add some restrictions. Hmm. I see. You can't filter non playable lands before choice cause some of the choice can allow to play next land and vice versa.

> 2. I copied this flow from [[Arcane Bombardment]]

If it depends on choice order. If it like hideaway code then yes -- it must use be used same way (allow to choose all and filter non-playable before cast)

> 'playOne' which sets the max targets to 1 on the TargetCard

Yes, it's fine -- if it allow to play only one card from a list then allow to choose only one. If player select non playable card/land then it's a problem of the player (it's optional effect).

Reply:
Sounds good. Updated the existing tests, wrote new tests for 607.3 playing multiple cards in chosen order and playing one or more lands. Updated Arcane Bombardment to not ask if you want to play the cards you already chose you want to play. Also added the optional boolean for Evercoat Ursine support in the future ("if there are cards exiled with it, you may play one of them without paying its mana cost.")